<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

  <title>HiGlass Presentation</title>

  <link rel="stylesheet" href="css/reveal.css">
  <link rel="stylesheet" href="css/theme/presentation.css">

  <!-- Theme used for syntax highlighting of code -->
  <link rel="stylesheet" href="lib/css/zenburn.css">

  <!-- Printing and PDF exports -->
  <script>
   var link = document.createElement( 'link' );
   link.rel = 'stylesheet';
   link.type = 'text/css';
   link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
   document.getElementsByTagName( 'head' )[0].appendChild( link );
</script>
<script src="lib/js/head.min.js"></script>
<script src="js/reveal.js"></script>
</head>
<body>
  <div class="reveal">
   <div class="slides">
    <section>
        <h2>Genomes, maps and matrices</h2> 
        <h3><i>How we use the technology behind web-based maps to explore trillion-cell genome contact matrices</i></h3>
        <br>
        <p>
            Peter Kerpedjiev<br>
            Postdoc, Gehlenborg Lab<br>
            Harvard Medical School
        </p>
    </section>

    <section> 
        <h3>The Question</h3>

        <p><i>What is the structure of DNA within the nucleus?</i></p>
    </section>

    <section>
        <h3>DNA</h3>
        <table width="1000px">
            <tr>
                <td style="vertical-align: middle; width: 400px; height: 400px">
                    <img src="img/DNA_simple2.svg" width="400px"/>
                </td>

                <td style="vertical-align: middle; width: 400px; height: 400px">
                    <img src="img/DNA-Chromosome.jpg" width="400px"/>
                </td>
            </tr>
        </table>

        <p><small>
        3.4e-10m distance between base pairs * 6e9 base pairs ~= 2.04 meters of DNA in a single cell
        </small></p>
    </section>

    <section>
        <h3>The nucleus</h3>

        <img src="img/nucleusfigure1.jpg" width="400px"/>

        <p><small>
            Nucleus is about 1 micrometer (1e-6 meters) in diameter.
        </small></p>
    </section>

    <section>
        <h3>DNA is tightly packed within the nucleus</h3>

        <table >
            <tr>
                <td style="vertical-align: middle; width: 400px; height: 400px">
                    <img src="img/uu-to-antelope.png" width="400px"/>
                </td>

                <td style="text-align: center; vertical-align: middle; height: 400px; width: 300px">
                    <img src="img/Basketball.png" width="200px"/>
                </td>
            </tr>
        </table>

    </section>

    <section>
        <h3>Why do we care?</h3>
        <div style="display: flex; justify-content: space-evenly; margin-top: 50px;">
            <img class="fragment" src="img/equilibrium-globule.png">
            <img class="fragment" src="img/fractal-globule.png">
        </div>
        <p class='fragment'><small>It affects biological processes such as gene expression (Zuin et al. 2014), cancer (Hnisz et al. 2016), X-Chromosome inactivation (Chen et al. 2014) </small></p>
    </section>

    <section>
        <h3>How do we measure DNA structure?</h3>
        <div class="fragment" style="display: flex; align-items: center; justify-content: space-evenly" >
            <div>
                <h4>X-Ray Crystallography?</h4>
                <br>
                <p><small>Structure 5I6Q from PDB</small></p>
            </div>
            <img src="img/5I6Q.png" width="600"/>
        </div>
        <aside class="notes" src="This is how the structure of DNA was initially determined" />
    </section>

    <section>
        <h3>How do we measure DNA structure?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly" >
            <div>
                <h4>Microscopy?</h4>
                <br>
                <p><small>Horng et al. Science 2017</small></p>
            </div>

            <img src="img/F6.large.jpg" width="500"/>
        </div>
        <aside class="notes" src="There are, of course, other types of microscopy too." />
        <div class="footer">
        </div>
    </section>

    <section>
        <h3>How do we measure DNA structure?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly" >

            <h4>Hi-C?</h4>

            <img src="img/higlass-sequencing.png" width="500"/>
        </div>
        <aside class="notes" src="" />
    </section>

    <section>
        <h3>What is Hi-C data?</h3>
        Hi-C data describes contacts between distal portions of DNA
        <table>
            <tr>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-1.png" height="26%" class="no-border"></td>
                </td>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-2.png" height="26%" class="no-border"></td>
                </td>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-3.png" height="26%" class="no-border"></td>
                </td>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-4.png" height="26%" class="no-border"></td>
                </td>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-5.png" height="26%" class="no-border"></td>
                </td>
                <td class="fragment no-border align-top">
                    <img src="img/aiden-et-al-6.png" height="26%" class="no-border"></td>
                </td>
            </tr>
        </table>

        <div class="footer fragment">
            <p>Lieberman-Aiden, Erez, et al. Science 326.5950 (2009): 289-293.</p>
        </div>
    </section>

    <section>
        <h3>Hi-C Data (Overview)</h3>

        <aside class="notes">
            <ul>
                <li>The point of Hi-C data is to show the 3D conformation of the genome.</li> 
                <li>In an ideal world, we'd use something like X-ray crystallography or NMR or cryo-EM to get a snapshot.</li>
                <li>In the real world, we make use of the technology that we have (sequencing)</li>
            </ul>
        </aside>


        </aside>

        <table>
            <tr>
                <td class="fragment no-border">
                    <img src="img/hic-data-fritz-1.png" height="40%" class="no-border"></td>
                </td>
                <td class="fragment no-border">
                    <img src="img/hic-data-fritz-2.png" height="40%" class="no-border"></td>
                </td>
                <td class="fragment no-border">
                    <img src="img/hic-data-fritz-3.png" height="40%" class="no-border"></td>
                </td>
                <td class="fragment no-border">
                    <img src="img/hic-data-fritz-4.png" height="40%" class="no-border"></td>
                </td>
                <td class="fragment no-border">
                    <img src="img/hic-data-fritz-5.png" height="40%" class="no-border"></td>
                </td>
            </tr>
        </table>
        <p class="fragment">The end result of a Hi-C assay is a <b>sparse 3 million by 3 million contact matrix.</b></p>
        <div class="footer">
            <p>Figure from F Lekschas et al.  IEEE Trans Vis Comput Graph. 2017 Aug 29.</p>
        </div>
    </section>

    <section>
        <h3>Hi-C Contact Matrices</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly" >

            <p class="fragment" data-fragment-index="2">To understand the structure of DNA within the nucleus, we need
            to be able to <b>explore</b> and <b>compare</b> Hi-C contact
            matrices in the <b>context</b> of other genomic data and annotations</p>

            <img src="img/higlass-sequencing.png" width="500" class="fragment" data-fragment-index="1"/>
        </div>
        <aside class="notes" src="" />
    </section>

    <section>
        <h3>Goal of this project</h3>

        <p>Create software to explore, compare and contextualize Hi-C contact matrices</p>
    </section>


    <section>
        <h3>Our Challenges</h3>
    </section>

    <section>
        <h3>Challenge #1</h3>
        <p class="fragment">How do we explore a 3M x 3M matrix in a web browser?</p>

        <ul>
            <li class="fragment">9 trillion cells</li>
            <li class="fragment">22Gb of compressed data</li>
            <li class="fragment">Dynamic range spanning multiple orders of magnitude</li>
        </ul>

    </section>

    <section>
        <h3>Solution</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 500px" class="fragment">
                <p>
                    <b>Solution:</b> Allow the user to navigate the matrix like a map: zooming and panning to regions of interest.
                </p>
            </div>
            <iframe src="hg-demo4.html" height="500px" width="500px" class="fragment"></iframe>
        </div>
    </section>

    <section>
        <h3>How does this work?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div>
                <img class='fragment' src="img/higlass-client-server.png" width="400px" data-fragment-index="2"/><br>
                <img class='fragment' src="img/higlass-client-server-2.png" width="400px" data-fragment-index="3"/>
            </div>
            <iframe src="hg-demo5.html" height="500px" width="500px" class="fragment" data-fragment-index="1"></iframe>
        </div>
    </section>

    <section>
        <h3>How is the client built?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 150px; margin-right: 30px">
                <img src="img/higlass-client.png" width="80px" />
            </div>
            <table style="vertical-align: middle">
                <tr>
                    <td>
                        <p><small>Fast WebGL-based rendering</small></p>
                    </td>
                    <td style="vertical-align: middle">
                        <img src="img/pixijs.png" width="160px" />
                    </td>
                </tr>
                <tr>
                    <td>
                        <p><small>Scales, utility functions and behaviors such as zooming and brushing</small></p>
                    </td>
                    <td style="vertical-align: middle">
                        <img src="img/d3.png" width="60px" />
                    </td>
                </tr>
                <tr>
                    <td>
                        <p><small>User interface composition and manipulation</small></p>
                    </td>
                    <td style="vertical-align: middle">
                        <img src="img/react.png" width="250px" />
                    </td>
                </tr>
            </table>
        </div>
    </section>

    <section>
        <h3>How is the server built?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 150px; margin-right: 30px">
                <img src="img/higlass-server.png" width="80px" />
            </div>
            <table style="vertical-align: middle">
                <tr>
                    <td>
                        <p><small>Content management</small></p>
                    </td>
                    <td style="vertical-align: middle">
                        <img src="img/django.png" width="160px" />
                    </td>
                </tr>
                <tr>
                    <td rowspan="2" style="vertical-align: middle">
                        <p><small>Data storage</small></p>
                    </td>
                    <td style="vertical-align: middle">
                        <img src="img/hdf.png" width="60px" />
                    </td>
                </tr>
                <tr>
                    <td style="vertical-align: middle">
                        <img src="img/cooler-logo.png" width="150px" />
                    </td>
                </tr>
            </table>
        </div>
    </section>

    <section>
        <h3>How are they deployed?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <img src="img/higlass-docker-composition.png" width="500px" />
        </div>
        <p class="fragment"><small>The HiGlass Docker container can either be deployed locally or on cloud infrastructure (eg. http://higlass.io).</small></p>
    </section>

    <section>
        <h3>How are the components tested?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 150px; margin-right: 30px">
                <img src="img/higlass-client.png" width="80px" />
            </div>
            <div>
              <video width="500" muted="muted" autoplay controls loop preload="none">
                  <source src="img/higlass_tests.mp4" type="video/mp4">
              </video>
            </div>
        </div>
        Integration testing to verify functionality of UI and visualization elements
    </section>

    <section>
        <h3>How are the components tested?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 150px; margin-right: 30px">
                <img src="img/higlass-server.png" width="80px" />
            </div>
            <div>
                <img src="img/test-higlass-server.png" width="700px" />
            </div>
        </div>
        Unit tests for models and APIs.
    </section>

    <section>
        <h3>How are the components tested?</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 150px; margin-right: 30px">
                <img src="img/higlass-docker.png" width="80px" />
            </div>
            <div>
                <img src="img/test-higlass-docker.png" width="700px" />
            </div>
        </div>
        Use Travis CI to build Docker container and call REST API endpoints.
    </section>

    <section>
        <h3>Open Questions</h3>
        <ul>
            <li>How do we efficiently retrieve subsections of a compressed sparse trillion cell matrix?
                <small>
                <ul>
                    <li>E.g. extract a 256x256 submatrix in less than 100ms</li>
                </ul>
                </small>
            </li>
            <li>How do we render millions of data points at >30 fps?</li>
            <li>What are the minimum APIs required to request variations of the data</li> 
                <small>
                <ul>
                    <li>E.g. Give me dataset x divided by dataset y</li>
                </ul>
                </small>
            </li>
        </ul>
    </section>


    <!--
        <section>
            <h3>Big data: how do we display a 3M x 3M matrix in a browser?</h3>
            <div class="fragment">
                <img src="img/aggregation.png" width="500"/>
                <p><b>Aggregation! Average adjacent cells in the matrix to create lower resolutions.</b></p>
            </div>
        </section>
    -->

    <section>
        <h3>Challenge #2</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 500px" class="fragment">
                <p>
                    How do we place the Hi-C contact matrix in the context of other genomic data?
                </p>
            </div>
            <iframe src="hg-demo4.html" height="500px" width="500px" class="fragment"></iframe>
        </div>
    </section>

    <section>
        <h3>Solution</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 500px">
                Add other genomic data as "tracks" along the axes.
            </div>
            <iframe src="hg_demo1.html" height="500" width="500"></iframe>

        </div>
    </section>

    <section>
        <h3> How do we do this? </h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div>
                <img class='fragment' src="img/higlass-client-server-annotations.png" width="400px" data-fragment-index="2"/><br>
                <img class='fragment' src="img/higlass-client-server-annotations-2.png" width="400px" data-fragment-index="3"/>
            </div>
            <iframe src="hg-demo6.html" height="300px" width="500px" class="fragment" data-fragment-index="1"></iframe>
        </div>
    </section>

    <section>
        <h3>What data do we display as 1D tracks?</h3>

        <table>
            <tr class='fragment'>
                <td style="vertical-align: middle">
                    Chromosome boundaries
                </td>
                <td style="vertical-align: middle">
                    <img src="img/chromosome-axis.png" height="100px"/>
                </td>
                <td style="vertical-align: middle; width:350px">
                    <small>
                        chromSizes file
                    </small>
                </td>
            </tr>

            <tr class='fragment'>
                <td style="vertical-align: middle">
                    Gene Annotations
                </td>
                <td style="vertical-align: middle">
                    <img src="img/higlass-gene-annotations.png" height="100px"/>
                </td>
                <td style="vertical-align: middle; width:350px">
                    <small>
                        SQLite (beddb)
                    </small>
                </td>
            </tr>

            <tr class='fragment'>
                <td style="vertical-align: middle">
                    Signal tracks (e.g. TF binding)
                </td>
                <td style="vertical-align: middle">
                    <img src="img/higlass-line-track.png" height="100px"/>
                </td>
                <td style="vertical-align: middle; width:350px">
                    <small>
                        bigWig, hitile
                    </small>
                </td>
            </tr>
        </table>
    </section>

    <section>
        <h3> Gene Annotations</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div style="width: 500px">
                <small>
                <ul>
                    <li class="fragment"><b>uuid:</b> unique identifier</li>
                    <li class="fragment"><b>startPos:</b> start of annotation</li>
                    <li class="fragment"><b>endPos:</b> end of annotations</li>
                    <li class="fragment"><b>chrOffset:</b> offset from the start of the containing chromosome</li>
                    <li class="fragment"><b>importance:</b> occlusion order based on number of citations</li>
                    <li class="fragment"><b>fields:</b> annotations's data (e.g. exon positions)</li>
                </ul>
                </small>
            </div>
            <iframe src="hg-demo3.html" height="300" width="500"></iframe>

        </div>
    </section>

    <section>
        <h3>Challenge #3</h3>
            <p> How do we compare data from multiple experimental conditions to
            gain insights and communicate findings?  </p>
        </h3>
    </section>

    <section>
        <h3>Solution</h3>
            <iframe src="hg_demo2.html" height="400" width="960"></iframe>
            <p>Create a flexible framework for synchronizing zoom and location across views.</p>
    </section>

    <section>
        <h3>Solution</h3>

            <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
                <img src="img/higlass-interaction-multiple-views-1.png" height="200px"/>
                <img src="img/higlass-interaction-multiple-views-2.png" width="300px"/>
            </div>
            <p> Views can be arbitrarily linked according to their zoom and location. They can also "project" their viewport onto another view to provide a context and detail.</p>
    </section>

    <section>
        <h3>Solution</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div>View Linking Configuration<br>
            <img src="img/higlass-interaction-multiple-views-2.png" width="300px"/>
            </div>
            <iframe src="hg-demo8.html" height="500" width="500"></iframe>
        </div>
    </section>

    <section>
        <h3>Summary</h3>
        <p>We created software to explore, compare and contextualize Hi-C contact matrices</p>
        <p>It's available at <a href="http://higlass.io">http://higlass.io</a></p>
        <h3 class='fragment'>Now what?</h3>
    </section>

    <section>
        <h3>Custom Data Sources</h3>
        <div style="display: flex; align-items: center; justify-content: space-evenly; margin-top: 50px;">
            <div>
                <h4 class='fragment' data-fragment-index="2">Option #1</h4>
                <img class='fragment' src="img/higlass-client-server.png" width="400px" data-fragment-index="2"/><br>
                <img class='fragment' src="img/higlass-client-server-2.png" width="400px" data-fragment-index="3"/>
            </div>
            <div>
                <h4 class='fragment' data-fragment-index="4">Option #2</h4>
                <img class='fragment' src="img/higlass-client-notebook.png" width="400px" data-fragment-index="4"/><br>
                <img class='fragment' src="img/higlass-client-notebook-2.png" width="400px" data-fragment-index="5"/>
            </div>
        </div>
    </section>

    <section>
        <h3>Example - HiGlass Running in Jupyter</h3>
        <img src="img/jupyter-higlass.png" width="600px" />
    </section>

    <section>
        <h3>Custom Data Types - GeoJSON</h3>
        <img src="img/ski-areas-in-france.gif" /><br/>
        <small><a href="https://github.com/flekschas/higlass-geojson">https://github.com/flekschas/higlass-geojson</a></small>

    </section>

    <section>
        <h3>Marketing</h3>
        <p>HiGlass is/was used by:</p>
        <ul>
            <li class="fragment">Over 30 academic laboratories</li>
            <li class="fragment">4 companies (therapeutics, diagnostics, and space propulsion)</li>
            <li class="fragment">1 diagnostic laboratory (non-profit)</li>
        </ul>
    </section>

    <section>
        <h3>Acknowledgements</h3>
        <div style="display: flex; justify-content: space-between;">
            <div style="width: 500px">
            <p><b><u>Harvard Medical School</u></b></p>
            <ul>
                <li>Nils Gehlenborg</li>
                <li>Fritz Lekschas (also at John Paulson School of Engineering)</li>
                <li>Chuck McCallum</li>
                <li>4DN DCIC Team</li>
            </ul>

            </div>
            <div>
            <p><b><u>MIT</u></b></p>
            <ul>
                <li>Nezar Abdennur</li>
                <li>Leonid Mirny</li>
            </ul>
        </div>
        </div>
    </section>
</div>

</div>


<script>
			// More info about config & dependencies:
			// - https://github.com/hakimel/reveal.js#configuration
			// - https://github.com/hakimel/reveal.js#dependencies
			Reveal.initialize({
                height: 800,
                history: true,
                "transition": "none",
                dependencies: [
                { src: 'plugin/markdown/marked.js' },
                { src: 'plugin/markdown/markdown.js' },
                { src: 'plugin/notes/notes.js', async: true },
                { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } }
                ],

                minScale: 1,
                maxScale: 1
            });
        </script>
    </body>

    </html>
